### PR TITLE
Issue 214: Custom family prediction functions for lognormal

### LINKED ISSUE #214: Posterior prediction for lognormal
The code `prep <- brms::prepare_predictions(fit)` works.

To fit into the way `brms` handles things what we would need to do is write custom functions like:

```
log_lik_latent_lognormal <- function(i, prep) {
}
```

```
posterior_predict_latent_lognormal <- function(i, prep, ...) {
}
```

```
posterior_epred_latent_lognormal <- function(prep) {
}
```

With these three things defined then all the standard `brms` post-processing should work.

Downside: have to write these functions by family class.

Suggestion: write this strata prediction function still (in a way which is 100% class agnostic) and then also write these above functions for the most common families (could just start with lognormal).

I'd say that these functions are more likely to be used by expert users of the package, and then the `delay_samples` function would be more likely to be used for "off the shelf" analysis by most users.

_Originally posted by @athowes in https://github.com/epinowcast/epidist/issues/188#issuecomment-2260072005_

### PR DESCRIPTION
**Pull request by athowes**
## Description

This PR closes #214.

## Checklist

- [x] My PR is based on a package issue and I have explicitly linked it.
- [x] I have included the target issue or issues in the PR title in the for Issue(s) *issue-numbers*: PR title
- [x] I have read the [contribution guidelines](https://github.com/epinowcast/.github/blob/main/CONTRIBUTING.md).
- [x] I have tested my changes locally.
- [x] I have added or updated unit tests where necessary.
- [x] I have updated the documentation if required.
- [x] My code follows the established coding standards.
- [ ] I have added a news item linked to this PR.
- [x] I have reviewed CI checks for this PR and addressed them as far as I am able.

### CODE REVIEW

**Comment by athowes** [comment]
Useful reference for implementing predictions:

```c
// generated with brms 2.21.6
functions {
  // code chunks used from epidist 0.0.0.9000

  // Here the strings
// * lognormal
// * vector mu, vector sigma
// * mu, sigma
// are/have been replaced using regex

real latent_lognormal_lpdf(vector y, vector mu, vector sigma, vector pwindow,
                           vector swindow, array[] real obs_t) {
  int n = num_elements(y);
  vector[n] d = y - pwindow + swindow;
  vector[n] obs_time = to_vector(obs_t) - pwindow;
  return lognormal_lpdf(d | mu, sigma) - lognormal_lcdf(obs_time | mu, sigma);
}
}
data {
  int<lower=1> N;  // total number of observations
  vector[N] Y;  // response variable
  // data for custom real vectors
  array[N] real vreal1;
  // data for custom real vectors
  array[N] real vreal2;
  // data for custom real vectors
  array[N] real vreal3;
  int prior_only;  // should the likelihood be ignored?
  int wN;
  array[N - wN] int noverlap;
  array[wN] int woverlap;
}
transformed data {
}
parameters {
  real Intercept;  // temporary intercept for centered predictors
  real Intercept_sigma;  // temporary intercept for centered predictors
  vector<lower = 0, upper = 1>[N] swindow_raw;
vector<lower = 0, upper = 1>[N] pwindow_raw;
}
transformed parameters {
  real lprior = 0;  // prior contributions to the log posterior
  vector<lower = 0>[N] pwindow;
vector<lower = 0>[N] swindow;
swindow = to_vector(vreal3) .* swindow_raw;
pwindow[noverlap] = to_vector(vreal2[noverlap]) .* pwindow_raw[noverlap];
if (wN) {
  pwindow[woverlap] = swindow[woverlap] .* pwindow_raw[woverlap];
}
  lprior += normal_lpdf(Intercept | 2, 0.5);
  lprior += normal_lpdf(Intercept_sigma | 0, 0.5);
}
model {
  swindow_raw ~ uniform(0, 1);
pwindow_raw ~ uniform(0, 1);
  // likelihood including constants
  if (!prior_only) {
    // initialize linear predictor term
    vector[N] mu = rep_vector(0.0, N);
    // initialize linear predictor term
    vector[N] sigma = rep_vector(0.0, N);
    mu += Intercept;
    sigma += Intercept_sigma;
    sigma = exp(sigma);
    target += latent_lognormal_lpdf(Y | mu, sigma, pwindow, swindow, vreal1);
  }
  // priors including constants
  target += lprior;
}
generated quantities {
  // actual population-level intercept
  real b_Intercept = Intercept;
  // actual population-level intercept
  real b_sigma_Intercept = Intercept_sigma;
}
```

**Comment by athowes** [comment]
Remains to do:

* [ ] Make sure `posterior_predict_latent_lognormal` works as expected
  * How to get it to return integers if required?
* [ ] Add unit tests for these
  * Feels difficult to arrive at suitable `prep` without fitting the model (i.e. closer to an integration test)
* [ ] Maybe delete the `latent_lognormal_lpdf`? It isn't needed for now. It'll be useful for #240

File: R/latent_lognormal.R
Comment (athowes) [inline R/latent_lognormal.R:41]: Think more about this (?)

Comment (athowes) [reply R/latent_lognormal.R:41]: Look at `simulate_double_censored`

Comment (athowes) [reply R/latent_lognormal.R:41]: `vreal2 <- prep$data$vreal2[i]` contains the information about the size of the window

File: R/latent_lognormal.R
Comment (athowes) [inline R/latent_lognormal.R:21]: Don't need this

File: tests/testthat/test-unit-predict.R
Comment (athowes) [inline tests/testthat/test-unit-predict.R:2]: I don't want to have to run this every time, but I want to always have `prep` available for these tests. What to do? Can't have it here because it'll error on CRAN

Comment (seabbs) [reply tests/testthat/test-unit-predict.R:2]: In epinowcast we store an example in inst and use that which is slightly annoying but works.

Could have it in setup.R and have these tests only run if not on CRAN

Look at what brms does?

File: R/utils.R
Comment (seabbs) [inline R/utils.R:34]: do we want to export this? Arg for is useful for people to. have to construct there own thing. Arg against is that is mainly developers and its more stuff to support

**Review by seabbs — COMMENTED** [review]
This all looks pretty good to me

**Comment by seabbs** [comment]
> Error: Error in x[i, , drop = FALSE] : subscript out of bounds

I see this error in the epidist tests which seems odd (rerunning to see if it goes away). My expectation is that it is unrelated to this PR.

File: R/utils.R
Comment (athowes) [reply R/utils.R:34]: Am fine with not exporting it, will make that change

File: tests/testthat/test-unit-predict.R
Comment (athowes) [reply tests/testthat/test-unit-predict.R:2]: It has `brms:::brmsfit_example1` (so putting examples as unexported data). It also constructs the `prep` object including using functions like `brms:::add_dummy_draws`

Comment (athowes) [reply tests/testthat/test-unit-predict.R:2]: Maybe I'm wrong that this means putting examples in unexported data

Comment (athowes) [reply tests/testthat/test-unit-predict.R]: I've done the `inst/extdata` approach for now but I think this needs further thought. We have an issue for "reduce number of times model is fit in tests"

**Review by seabbs — APPROVED** [review]
Nice!

**Review by seabbs — APPROVED** [review]

**Comment by athowes** [comment]
@seabbs I'm not sure why it's failing `R-CMD-check` on Windows latest but think that it'd be OK to merge it. If you agree could you overwrite the required status and merge?

**Comment by athowes** [comment]
The issue with some `R-CMD-check` is "Assertion on 'files' failed: File does not exist" to do with `cmdstanr`. It seems to be something like failing to cache models properly. It comes up in various places related to `cmdstanr`. Unsure how to fix at the moment. Perhaps some change to the actions code?

* https://github.com/stan-dev/cmdstanr/issues/347
* https://discourse.mc-stan.org/t/error-in-read-cmdstan-csv-assertion-on-files-failed/30150}
* https://discourse.mc-stan.org/t/error-in-cmdstanr-read-cmdstan-csv-for-mixed-model-win-10/23392

**Comment by athowes** [comment]
Well that's bad that what it took to pass these tests was to change `seed = 1` to `seed = 2` in the couple of places that failed but oh well..

**Review by kgostic — APPROVED** [review]
LGTM

**Review by kgostic — APPROVED** [review]
Re-approving after merge edit

**Review by kgostic — COMMENTED** [review]
I'd ask Zack -- I think we've encountered this problem elsewhere with CI.

**Pull request merged**
